plugins/discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java

```diff
@@ -57,15 +59,32 @@ public synchronized AmazonEC2 client() {
             return client;
         }
 
-        this.client = new AmazonEC2Client(buildCredentials(logger, settings), buildConfiguration(logger, settings));
+        AmazonEC2ClientBuilder builder = AmazonEC2ClientBuilder.standard()
+            .withCredentials(buildCredentials(logger, settings))
+            .withClientConfiguration(buildConfiguration(logger, settings));
+
         String endpoint = findEndpoint(logger, settings);
         if (endpoint != null) {
-            client.setEndpoint(endpoint);
+            builder.setEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(endpoint, buildRegion(endpoint)));
+        } else {
+            logger.debug("No endpoint defined. Using default endpoint [ec2.us-east-1.amazonaws.com].");
         }
 
+        this.client = builder.build();
+
         return this.client;
     }
 
+    private static String buildRegion(String endpoint) {
+        // We try to get the region from the metadata instance information
+        String region = new InstanceMetadataRegionProvider().getRegion();
+        if (region == null) {
+            // Or we try to get it from the endpoint itself
+            region = AwsHostNameUtils.parseRegion(endpoint, null);
```

Comment:
Don't we need to check for null again here? `pareRegion` can return null if the endpoint is "non standard". I think we need to throw an error in that case or the endpoint configuration will be incorrect (endpoint but null region)?

Reply:
Ah! Good catch. Indeed the AWS implementation returns null. I'm going to add this. Thanks